// contentful-management/dist/es-modules/adapters/REST/endpoints/entry.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import copy from 'fast-copy';
import * as raw from './raw';
import * as releaseEntry from './release-entry';
import { normalizeSelect } from './utils';
export const get = (http, params, rawData, headers) => {
  if (params.releaseId) {
    return releaseEntry.get(http, params);
  }
  return raw.get(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, {
    params: normalizeSelect(params.query),
    headers: _objectSpread({}, headers)
  });
};
export const getPublished = (http, params, rawData, headers) => {
  return raw.get(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/public/entries`, {
    params: normalizeSelect(params.query),
    headers: _objectSpread({}, headers)
  });
};
export const getMany = (http, params, rawData, headers) => {
  if (params.releaseId) {
    return releaseEntry.getMany(http, params);
  }
  return raw.get(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries`, {
    params: normalizeSelect(params.query),
    headers: _objectSpread({}, headers)
  });
};
export const patch = (http, params, data, headers) => {
  if (params.releaseId) {
    return releaseEntry.patch(http, params, data, headers !== null && headers !== void 0 ? headers : {});
  }
  return raw.patch(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, data, {
    headers: _objectSpread({
      'X-Contentful-Version': params.version,
      'Content-Type': 'application/json-patch+json'
    }, headers)
  });
};
export const update = (http, params, rawData, headers) => {
  var _rawData$sys$version;
  if (params.releaseId) {
    return releaseEntry.update(http, params, rawData, headers !== null && headers !== void 0 ? headers : {});
  }
  const data = copy(rawData);
  delete data.sys;
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, data, {
    headers: _objectSpread({
      'X-Contentful-Version': (_rawData$sys$version = rawData.sys.version) !== null && _rawData$sys$version !== void 0 ? _rawData$sys$version : 0
    }, headers)
  });
};
export const del = (http, params) => {
  return raw.del(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`);
};
export const publish = (http, params, rawData) => {
  var _params$locales;
  const payload = (_params$locales = params.locales) !== null && _params$locales !== void 0 && _params$locales.length ? {
    add: {
      fields: {
        '*': params.locales
      }
    }
  } : null;
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/published`, payload, {
    headers: {
      'X-Contentful-Version': rawData.sys.version
    }
  });
};
export const unpublish = (http, params, rawData) => {
  var _params$locales2;
  if ((_params$locales2 = params.locales) !== null && _params$locales2 !== void 0 && _params$locales2.length) {
    const payload = {
      remove: {
        fields: {
          '*': params.locales
        }
      }
    };
    return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/published`, payload, {
      headers: {
        'X-Contentful-Version': rawData === null || rawData === void 0 ? void 0 : rawData.sys.version
      }
    });
  } else {
    return raw.del(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/published`);
  }
};
export const archive = (http, params) => {
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/archived`);
};
export const unarchive = (http, params) => {
  return raw.del(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/archived`);
};
export const create = (http, params, rawData) => {
  if (params.releaseId) {
    return releaseEntry.create(http, params, rawData, {});
  }
  const data = copy(rawData);
  return raw.post(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries`, data, {
    headers: {
      'X-Contentful-Content-Type': params.contentTypeId
    }
  });
};
export const createWithId = (http, params, rawData) => {
  if (params.releaseId) {
    return releaseEntry.createWithId(http, params, rawData, {});
  }
  const data = copy(rawData);
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, data, {
    headers: {
      'X-Contentful-Content-Type': params.contentTypeId
    }
  });
};
export const references = (http, params) => {
  const {
    spaceId,
    environmentId,
    entryId,
    include
  } = params;
  const level = include || 2;
  return raw.get(http, `/spaces/${spaceId}/environments/${environmentId}/entries/${entryId}/references?include=${level}`);
};
