// contentful-management/dist/es-modules/adapters/REST/endpoints/webhook.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import copy from 'fast-copy';
import * as raw from './raw';
import { normalizeSelect } from './utils';
const getBaseUrl = params => `/spaces/${params.spaceId}/webhook_definitions`;
const getWebhookCallBaseUrl = params => `/spaces/${params.spaceId}/webhooks`;
const getWebhookUrl = params => `${getBaseUrl(params)}/${params.webhookDefinitionId}`;
const getWebhookCallUrl = params => `${getWebhookCallBaseUrl(params)}/${params.webhookDefinitionId}/calls`;
const getWebhookCallDetailsUrl = params => `${getWebhookCallBaseUrl(params)}/${params.webhookDefinitionId}/calls/${params.callId}`;
const getWebhookHealthUrl = params => `${getWebhookCallBaseUrl(params)}/${params.webhookDefinitionId}/health`;
const getWebhookSettingsUrl = params => `/spaces/${params.spaceId}/webhook_settings`;
const getWebhookSigningSecretUrl = params => `${getWebhookSettingsUrl(params)}/signing_secret`;
const getWebhookRetryPolicyUrl = params => `${getWebhookSettingsUrl(params)}/retry_policy`;
export const get = (http, params) => {
  return raw.get(http, getWebhookUrl(params));
};
export const getManyCallDetails = (http, params) => {
  return raw.get(http, getWebhookCallUrl(params), {
    params: normalizeSelect(params.query)
  });
};
export const getCallDetails = (http, params) => {
  return raw.get(http, getWebhookCallDetailsUrl(params));
};
export const getHealthStatus = (http, params) => {
  return raw.get(http, getWebhookHealthUrl(params));
};
export const getMany = (http, params) => {
  return raw.get(http, getBaseUrl(params), {
    params: normalizeSelect(params.query)
  });
};
export const getSigningSecret = (http, params) => {
  return raw.get(http, getWebhookSigningSecretUrl(params));
};

/**
 * @deprecated The EAP for this feature has ended. This method will be removed in the next major version.
 */
export const getRetryPolicy = (http, params) => {
  return raw.get(http, getWebhookRetryPolicyUrl(params));
};
export const create = (http, params, rawData, headers) => {
  const data = copy(rawData);
  return raw.post(http, getBaseUrl(params), data, {
    headers
  });
};
export const createWithId = (http, params, rawData, headers) => {
  const data = copy(rawData);
  return raw.put(http, getWebhookUrl(params), data, {
    headers
  });
};
export const update = async (http, params, rawData, headers) => {
  var _rawData$sys$version;
  const data = copy(rawData);
  delete data.sys;
  return raw.put(http, getWebhookUrl(params), data, {
    headers: _objectSpread({
      'X-Contentful-Version': (_rawData$sys$version = rawData.sys.version) !== null && _rawData$sys$version !== void 0 ? _rawData$sys$version : 0
    }, headers)
  });
};
export const upsertSigningSecret = async (http, params, rawData) => {
  const data = copy(rawData);
  return raw.put(http, getWebhookSigningSecretUrl(params), data);
};

/**
 * @deprecated The EAP for this feature has ended. This method will be removed in the next major version.
 */
export const upsertRetryPolicy = async (http, params, rawData) => {
  const data = copy(rawData);
  return raw.put(http, getWebhookRetryPolicyUrl(params), data);
};
export const del = (http, params) => {
  return raw.del(http, getWebhookUrl(params));
};
export const deleteSigningSecret = async (http, params) => {
  return raw.del(http, getWebhookSigningSecretUrl(params));
};

/**
 * @deprecated The EAP for this feature has ended. This method will be removed in the next major version.
 */
export const deleteRetryPolicy = async (http, params) => {
  return raw.del(http, getWebhookRetryPolicyUrl(params));
};
