// contentful-management/dist/es-modules/entities/workflow.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import copy from 'fast-copy';
import { wrapCollection } from '../common-utils';
import enhanceWithMethods from '../enhance-with-methods';
/**
 * @private
 */
export default function createWorkflowApi(makeRequest) {
  const getParams = Workflow => ({
    spaceId: Workflow.sys.space.sys.id,
    environmentId: Workflow.sys.environment.sys.id,
    workflowId: Workflow.sys.id
  });
  return {
    update: function () {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'update',
        params: getParams(raw),
        payload: raw
      }).then(data => wrapWorkflow(makeRequest, data));
    },
    get: function () {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'get',
        params: getParams(raw)
      }).then(data => wrapWorkflow(makeRequest, data));
    },
    delete: function () {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'delete',
        params: _objectSpread(_objectSpread({}, getParams(raw)), {}, {
          version: raw.sys.version
        })
      }).then(() => {
        // noop
      });
    },
    complete: function () {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'complete',
        params: _objectSpread(_objectSpread({}, getParams(raw)), {}, {
          version: raw.sys.version
        })
      }).then(() => {
        // noop
      });
    }
  };
}

/**
 * @private
 */
export function wrapWorkflow(makeRequest, data) {
  const Workflow = toPlainObject(copy(data));
  const WorkflowWithMethods = enhanceWithMethods(Workflow, createWorkflowApi(makeRequest));
  return freezeSys(WorkflowWithMethods);
}

/**
 * @private
 */
export const wrapWorkflowCollection = wrapCollection(wrapWorkflow);
