// @contentful/app-scripts/lib/clean-up/build-clean-up-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildCleanUpSettings = buildCleanUpSettings;
const inquirer_1 = require("inquirer");
const constants_1 = require("../constants");
const get_app_info_1 = require("../get-app-info");
async function buildCleanUpSettings(options) {
    const { keep, host } = options;
    const prompts = [];
    if (keep === undefined) {
        prompts.push({
            type: 'number',
            name: 'keep',
            message: `The amount of newest bundles to keep:`,
            default: constants_1.DEFAULT_BUNDLES_TO_KEEP,
        });
    }
    if (!host) {
        prompts.push({
            name: 'host',
            message: `Contentful CMA endpoint URL:`,
            default: constants_1.DEFAULT_CONTENTFUL_API_HOST,
        });
    }
    const { host: interactiveHost, ...appCleanUpSettings } = await (0, inquirer_1.prompt)(prompts);
    const hostValue = host || interactiveHost;
    const appInfo = await (0, get_app_info_1.getAppInfo)({ ...options, host: hostValue });
    return {
        keep: keep === undefined ? +appCleanUpSettings.keep : +keep,
        host: hostValue,
        ...appCleanUpSettings,
        ...appInfo,
    };
}
