// @contentful/app-scripts/lib/clean-up/get-clean-up-settings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCleanUpSettingsArgs = getCleanUpSettingsArgs;
const chalk_1 = __importDefault(require("chalk"));
const ora_1 = __importDefault(require("ora"));
const constants_1 = require("../constants");
const validate_arguments_1 = require("../validate-arguments");
const get_app_info_1 = require("../get-app-info");
const requiredOptions = {
    definitionId: '--definition-id',
    organizationId: '--organization-id',
    token: '--token',
};
async function getCleanUpSettingsArgs(options) {
    const validateSpinner = (0, ora_1.default)('Validating your input...').start();
    try {
        (0, validate_arguments_1.validateArguments)(requiredOptions, options, 'upload');
        const appInfo = await (0, get_app_info_1.getAppInfo)(options);
        return {
            ...appInfo,
            keep: options.keep !== undefined ? +options.keep : constants_1.DEFAULT_BUNDLES_TO_KEEP,
            host: options.host,
        };
    }
    catch (err) {
        console.log(`
      ${chalk_1.default.red('Validation failed!')}
      ${err.message}
    `);
        // eslint-disable-next-line no-process-exit
        process.exit(1);
    }
    finally {
        validateSpinner.stop();
    }
}
