// @contentful/app-scripts/lib/generate-function/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.warn = warn;
exports.error = error;
exports.wrapInBlanks = wrapInBlanks;
exports.highlight = highlight;
exports.choice = choice;
exports.success = success;
exports.code = code;
const chalk_1 = __importDefault(require("chalk"));
const types_1 = require("./types");
function warn(message) {
    console.log(`${chalk_1.default.yellow('Warning:')} ${message}`);
}
function error(message, error) {
    console.log(`${chalk_1.default.red('Error:')} ${message}`);
    if (error === undefined) {
        return;
    }
    else if (error instanceof types_1.InvalidTemplateError || error instanceof types_1.HTTPResponseError) {
        // for known errors, we just want to show the message
        console.log();
        console.log(error.message);
    }
    else if (error instanceof Error) {
        console.log();
        console.log(error);
    }
    else {
        const strigifiedError = String(error);
        console.log();
        console.log(`${strigifiedError.startsWith('Error: ') ? strigifiedError.substring(7) : strigifiedError}`);
    }
}
function wrapInBlanks(message) {
    console.log(' ');
    console.log(message);
    console.log(' ');
}
function highlight(str) {
    return chalk_1.default.bold(str);
}
function choice(str) {
    return chalk_1.default.cyan(str);
}
function success(str) {
    return chalk_1.default.greenBright(str);
}
function code(str) {
    return chalk_1.default.bold(str);
}
