// @contentful/app-scripts/lib/activate/get-activate-args.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getActivateSettingsArgs = getActivateSettingsArgs;
const chalk_1 = __importDefault(require("chalk"));
const ora_1 = __importDefault(require("ora"));
const get_app_info_1 = require("../get-app-info");
const validate_arguments_1 = require("../validate-arguments");
const requiredOptions = {
    organizationId: '--organization-id',
    definitionId: '--definition-id',
    bundleId: '--bundle-id',
    token: '--token',
};
async function getActivateSettingsArgs(options) {
    const validateSpinner = (0, ora_1.default)('Validating your input').start();
    try {
        (0, validate_arguments_1.validateArguments)(requiredOptions, options, 'activate');
        const appInfo = await (0, get_app_info_1.getAppInfo)(options);
        return {
            ...appInfo,
            bundleId: options.bundleId,
            host: options.host,
        };
    }
    catch (err) {
        console.log(`
      ${chalk_1.default.red('Validation failed')}
      ${err.message}
    `);
        // eslint-disable-next-line no-process-exit
        process.exit(1);
    }
    finally {
        validateSpinner.stop();
    }
}
