// @contentful/f36-core/dist/index.js
'use strict';

var F = require('react');
var core = require('@emotion/core');
var l = require('@contentful/f36-tokens');
var emotion = require('emotion');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var F__namespace = /*#__PURE__*/_interopNamespace(F);
var l__default = /*#__PURE__*/_interopDefault(l);

var ye=Object.defineProperty,ge=Object.defineProperties;var ue=Object.getOwnPropertyDescriptors;var w=Object.getOwnPropertySymbols;var W=Object.prototype.hasOwnProperty,H=Object.prototype.propertyIsEnumerable;var z=(e,t,o)=>t in e?ye(e,t,{enumerable:!0,configurable:!0,writable:!0,value:o}):e[t]=o,i=(e,t)=>{for(var o in t||(t={}))W.call(t,o)&&z(e,o,t[o]);if(w)for(var o of w(t))H.call(t,o)&&z(e,o,t[o]);return e},s=(e,t)=>ge(e,ue(t));var f=(e,t)=>{var o={};for(var r in e)W.call(e,r)&&t.indexOf(r)<0&&(o[r]=e[r]);if(e!=null&&w)for(var r of w(e))t.indexOf(r)<0&&H.call(e,r)&&(o[r]=e[r]);return o};var Pe=core.css`
  /* Remove default margin */
  /* Button has default margin in some browsers, like safari */
  body,
  h1,
  h2,
  h3,
  h4,
  p,
  figure,
  blockquote,
  dl,
  dd,
  button {
    margin: 0;
  }

  /* Remove list styles on ul, ol elements with a list role, which suggests default styling will be removed */
  ul[role='list'],
  ol[role='list'] {
    list-style: none;
  }

  /* Set core root defaults */
  html:focus-within {
    scroll-behavior: smooth;
  }

  /* Set core body defaults */
  body {
    text-rendering: optimizeSpeed;
    line-height: 1.5;
  }

  /* A elements that don't have a class get default styles */
  a:not([class]) {
    text-decoration-skip-ink: auto;
  }

  /* Make images easier to work with */
  img,
  picture {
    max-width: 100%;
    display: block;
  }

  /* Inherit fonts for inputs and buttons */
  input,
  button,
  textarea,
  select {
    font: inherit;
  }

  /* Remove all animations and transitions for people that prefer not to see them */
  @media (prefers-reduced-motion: reduce) {
    html:focus-within {
      scroll-behavior: auto;
    }
    *,
    *::before,
    *::after {
      animation-duration: 0.01ms !important;
      animation-iteration-count: 1 !important;
      transition-duration: 0.01ms !important;
      scroll-behavior: auto !important;
    }
  }
`,q=({withNormalize:e=!0,styles:t})=>F__namespace.createElement(core.Global,{styles:core.css`
        ${e?Pe:void 0};
        @font-face {
          font-family: 'Geist Mono';
          src: url('https://cdn.f36.contentful.com/fonts/geist-mono-1.4.01.woff2');
        }

        @font-face {
          font-family: 'Geist Sans';
          src: url('https://cdn.f36.contentful.com/fonts/geist-sans-1.4.01.woff2');
        }

        html {
          border: 0;
          box-sizing: border-box;
          margin: 0;
          padding: 0;
        }

        body {
          color: ${l__default.default.gray800};
          font-family: ${l__default.default.fontStackPrimary};
          font-size: ${l__default.default.fontSizeM};
          line-height: ${l__default.default.lineHeightM};
          font-synthesis-weight: none;
          /* Improves readability of lowercase L (l) and uppercase i (I) */
          font-feature-settings: 'ss05' on;
        }

        code {
          font-family: ${l__default.default.fontStackMonospace};
        }

        fieldset {
          border: 0;
          margin: 0;
          padding: 0;
        }

        *,
        *::after,
        *::before {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          box-sizing: inherit;
        }

        ${t};
      `});var Se={none:0,spacing2Xs:l__default.default.spacing2Xs,spacingXs:l__default.default.spacingXs,spacingS:l__default.default.spacingS,spacingM:l__default.default.spacingM,spacingL:l__default.default.spacingL,spacingXl:l__default.default.spacingXl,spacing2Xl:l__default.default.spacing2Xl,spacing3Xl:l__default.default.spacing3Xl,spacing4Xl:l__default.default.spacing4Xl};function c(e){return Se[e]}function V(e){return Object.assign({},e.margin?{margin:c(e.margin)}:null,e.marginLeft?{marginLeft:c(e.marginLeft)}:null,e.marginRight?{marginRight:c(e.marginRight)}:null,e.marginTop?{marginTop:c(e.marginTop)}:null,e.marginBottom?{marginBottom:c(e.marginBottom)}:null,e.padding?{padding:c(e.padding)}:null,e.paddingLeft?{paddingLeft:c(e.paddingLeft)}:null,e.paddingRight?{paddingRight:c(e.paddingRight)}:null,e.paddingTop?{paddingTop:c(e.paddingTop)}:null,e.paddingBottom?{paddingBottom:c(e.paddingBottom)}:null)}var Re="div";function g(e){let S=e,{display:t,className:o,margin:r,marginBottom:a,marginLeft:p,marginRight:n,marginTop:m,padding:y,paddingBottom:d,paddingLeft:u,paddingRight:x,paddingTop:P,testId:C,as:E=Re}=S,b=f(S,["display","className","margin","marginBottom","marginLeft","marginRight","marginTop","padding","paddingBottom","paddingLeft","paddingRight","paddingTop","testId","as"]);return {boxProps:i({className:emotion.cx(emotion.css({display:t}),emotion.css(i({},V({margin:r,marginBottom:a,marginLeft:p,marginRight:n,marginTop:m,padding:y,paddingBottom:d,paddingLeft:u,paddingRight:x,paddingTop:P}))),o),"data-test-id":C},b),Element:E}}function Y(e,t){let{boxProps:o,Element:r}=g(e);return F__namespace.createElement(r,s(i({},o),{ref:t}),e.children)}Y.displayName="Box";var Z=F__namespace.forwardRef(Y);var Te="div";function ee(N,U){var I=N,{isInline:e,alignItems:t,alignSelf:o,alignContent:r,flex:a,flexBasis:p,flexShrink:n,flexDirection:m,flexGrow:y,flexWrap:d,fullHeight:u,fullWidth:x,gap:P,justifyContent:C,justifyItems:E,justifySelf:b,order:T,children:S,as:k}=I,_=f(I,["isInline","alignItems","alignSelf","alignContent","flex","flexBasis","flexShrink","flexDirection","flexGrow","flexWrap","fullHeight","fullWidth","gap","justifyContent","justifyItems","justifySelf","order","children","as"]);var $;let{boxProps:G,Element:fe}=g(s(i({},_),{as:k||Te}));return F__namespace.createElement(fe,s(i({},G),{className:emotion.cx(emotion.css({display:e?"inline-flex":"flex",width:x?"100%":void 0,height:u?"100%":void 0,flex:a,flexBasis:p,flexShrink:n,flexDirection:m,gap:P==="none"?0:($=l__default.default[P])!=null?$:P,justifyContent:C,justifyItems:E,justifySelf:b,alignItems:t,alignSelf:o,alignContent:r,order:T,flexWrap:d,flexGrow:y}),G.className),ref:U}),S)}ee.displayName="Flex";var L=F__namespace.forwardRef(ee);var Ae="div";function te(T,b){var S=T,{alignContent:e,alignItems:t,children:o,columnGap:r="spacingM",columns:a="auto",flow:p,isInline:n,justifyContent:m,justifyItems:y,placeContent:d,placeItems:u,rowGap:x="none",rows:P="auto",as:C}=S,E=f(S,["alignContent","alignItems","children","columnGap","columns","flow","isInline","justifyContent","justifyItems","placeContent","placeItems","rowGap","rows","as"]);var N,I;let k=G=>typeof G=="number"?`repeat(${G}, minmax(0, 1fr))`:G,{boxProps:_,Element:U}=g(s(i({},E),{as:C||Ae}));return F__namespace.createElement(U,s(i({},_),{className:emotion.cx(emotion.css({display:n?"inline-grid":"grid",gridTemplateColumns:k(a),gridTemplateRows:k(P),flow:p,justifyContent:m,justifyItems:y,alignContent:e,alignItems:t,placeContent:d,placeItems:u,columnGap:(N=c(r))!=null?N:0,rowGap:(I=c(x))!=null?I:0}),_.className),ref:b}),o)}te.displayName="Grid";var oe=F__namespace.forwardRef(te);var we="div";function re(C,P){var E=C,{children:e,columnStart:t,columnEnd:o,rowStart:r,rowEnd:a,area:p,order:n,alignSelf:m,justifySelf:y,placeSelf:d,as:u}=E,x=f(E,["children","columnStart","columnEnd","rowStart","rowEnd","area","order","alignSelf","justifySelf","placeSelf","as"]);let b=p||[r||"auto",t||"auto",a||"auto",o||"auto"].join(" / "),{boxProps:T,Element:S}=g(s(i({},x),{as:u||we}));return F__namespace.createElement(S,s(i({},T),{className:emotion.cx(emotion.css({gridArea:b,order:n,alignSelf:m,justifySelf:y,placeSelf:d}),T.className),ref:P}),e)}re.displayName="GridItem";var v=F__namespace.forwardRef(re);var X=oe;X.Item=v;var Fe="div";function ne(y,m){var d=y,{flexDirection:e="row",alignItems:t="center",isInline:o=!1,spacing:r="spacingM",children:a,as:p}=d,n=f(d,["flexDirection","alignItems","isInline","spacing","children","as"]);let{boxProps:u,Element:x}=g(s(i({},n),{as:p||Fe}));return F__namespace.createElement(L,s(i({},u),{as:x,flexDirection:e,alignItems:t,isInline:o,className:u.className,gap:r,ref:m}),a)}ne.displayName="Stack";var ie=F__namespace.forwardRef(ne);function ae(){return {screenReaderOnly:emotion.css({position:"absolute",width:"1px",height:"1px",padding:0,margin:"-1px",overflow:"hidden",clip:"rect(0, 0, 0, 0)",border:0})}}var se=p=>{var n=p,{children:e,className:t,testId:o="cf-ui-ScreenReaderOnly",as:r="div"}=n,a=f(n,["children","className","testId","as"]);let m=ae();return F__namespace.default.createElement(r,s(i({},a),{"data-test-id":o,className:emotion.cx(m.screenReaderOnly,t)}),e)};function pe(e,t){return F.useMemo(()=>e||[t||"id",Math.round(Math.random()*1e4)].join("_"),[])}function me({isOpen:e,defaultIsOpen:t,onOpen:o,onClose:r}){let[a,p]=F.useState(t||!1),n=e!==void 0,m=n?e:a,y=F.useCallback(()=>{n||p(!1),r==null||r();},[n,r]),d=F.useCallback(()=>{n||p(!0),o==null||o();},[n,o]);return {isOpen:m,isControlled:n,handleClose:y,handleOpen:d}}function ce({onLoad:e}={}){let[t,o]=F.useState(!1),r=F.useRef(),a=F.useCallback(()=>{e==null||e(),o(!0);},[e]);return F.useEffect(()=>{r.current&&r.current.complete&&a();},[a]),{ref:r,loaded:t,onLoad:a}}function je(...e){return t=>{e.forEach(o=>{typeof o=="function"?o(t):o!=null&&(o.current=t);});}}var $e=({status:e})=>{switch(e){case"published":return {color:l__default.default.colorGreenBase,backgroundColor:l__default.default.colorGreenLightest};case"changed":return {color:l__default.default.colorBlueBase,backgroundColor:l__default.default.colorBlueLightest};case"deleted":case"archived":return {color:l__default.default.colorRedBase,backgroundColor:l__default.default.colorRedLightest};case"draft":return {color:l__default.default.colorOrangeDark,backgroundColor:"#ffefd5"};case"new":return {color:l__default.default.colorWhite,backgroundColor:l__default.default.blue500};default:return console.warn(`Unknown EntityStatus: ${e}`),{}}};var We={backgroundColor:l__default.default.gray200,"&:hover, &:active":{backgroundColor:l__default.default.gray200}},de={opacity:.5,cursor:"auto","&:hover":{backgroundColor:"unset"}},He=({isActive:e=!1,isDisabled:t=!1})=>emotion.cx(emotion.css({alignItems:"center",display:"flex",background:"none",borderWidth:"1px 0",borderColor:l__default.default.colorWhite,borderRadius:l__default.default.borderRadiusSmall,borderStyle:"solid",outline:"none",fontSize:l__default.default.fontSizeM,lineHeight:l__default.default.lineHeightM,fontWeight:l__default.default.fontWeightNormal,position:"relative",textAlign:"left",padding:`5px ${l__default.default.spacingXs}`,wordBreak:"break-word",whiteSpace:"break-spaces",cursor:"pointer",hyphens:"auto",minWidth:"150px",textDecoration:"none",color:l__default.default.gray700,"&:focus, &:hover":{backgroundColor:l__default.default.gray100},"&:focus":{boxShadow:`inset ${l__default.default.glowPrimary}`},"&:focus:not(:focus-visible)":{boxShadow:"unset"},"&:focus-visible":{boxShadow:`inset ${l__default.default.glowPrimary}`},"&:active":{color:l__default.default.gray700,backgroundColor:l__default.default.gray200},"&:disabled":de}),e&&emotion.css(We),t&&emotion.css(de));

exports.Box = Z;
exports.Flex = L;
exports.GlobalStyles = q;
exports.Grid = X;
exports.GridItem = v;
exports.ScreenReaderOnly = se;
exports.Stack = ie;
exports.getEntityStatusStyles = $e;
exports.getMenuItemStyles = He;
exports.mergeRefs = je;
exports.useBox = g;
exports.useControllableState = me;
exports.useId = pe;
exports.useImageLoaded = ce;
