// contentful-management/dist/es-modules/adapters/REST/endpoints/ai-action.js
const _excluded = ["sys"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
import copy from 'fast-copy';
import * as raw from './raw';
export const get = (http, params, headers) => {
  return raw.get(http, `/spaces/${params.spaceId}/ai/actions/${params.aiActionId}`, {
    headers
  });
};
export const getMany = (http, params, headers) => {
  return raw.get(http, `/spaces/${params.spaceId}/ai/actions`, {
    params: params.query,
    headers
  });
};
export const create = (http, params, data, headers) => {
  return raw.post(http, `/spaces/${params.spaceId}/ai/actions`, data, {
    headers
  });
};
export const update = (http, params, rawData, headers) => {
  var _rawData$sys$version;
  const data = copy(rawData);
  const {
      sys
    } = data,
    payload = _objectWithoutProperties(data, _excluded);
  return raw.put(http, `/spaces/${params.spaceId}/ai/actions/${params.aiActionId}`, payload, {
    headers: _objectSpread({
      'X-Contentful-Version': (_rawData$sys$version = rawData.sys.version) !== null && _rawData$sys$version !== void 0 ? _rawData$sys$version : 0
    }, headers)
  });
};
export const del = (http, params, headers) => {
  return raw.del(http, `/spaces/${params.spaceId}/ai/actions/${params.aiActionId}`, {
    headers
  });
};
export const publish = (http, params, rawData, headers) => {
  return raw.put(http, `/spaces/${params.spaceId}/ai/actions/${params.aiActionId}/published`, null, {
    headers: _objectSpread({
      'X-Contentful-Version': params.version
    }, headers)
  });
};
export const unpublish = (http, params, headers) => {
  return raw.del(http, `/spaces/${params.spaceId}/ai/actions/${params.aiActionId}/published`, {
    headers
  });
};
export const invoke = (http, params, data, headers) => {
  return raw.post(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/ai/actions/${params.aiActionId}/invoke`, data, {
    headers
  });
};
