// contentful-management/dist/es-modules/adapters/REST/endpoints/app-installation.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import * as raw from './raw';
import { normalizeSelect, normalizeSpaceId } from './utils';
import copy from 'fast-copy';
const getBaseUrl = params => `/spaces/${params.spaceId}/environments/${params.environmentId}/app_installations`;
const getBaseUrlForOrgInstallations = params => `/app_definitions/${params.appDefinitionId}/app_installations`;
export const getAppInstallationUrl = params => getBaseUrl(params) + `/${params.appDefinitionId}`;
export const get = (http, params) => {
  return raw.get(http, getAppInstallationUrl(params), {
    params: normalizeSelect(params.query)
  });
};
export const getMany = (http, params) => {
  return raw.get(http, getBaseUrl(params), {
    params: normalizeSelect(params.query)
  });
};
export const upsert = (http, params, rawData, headers) => {
  const data = copy(rawData);
  return raw.put(http, getAppInstallationUrl(params), data, {
    headers: _objectSpread(_objectSpread({}, headers), params.acceptAllTerms && {
      'X-Contentful-Marketplace': 'i-accept-end-user-license-agreement,i-accept-marketplace-terms-of-service,i-accept-privacy-policy'
    })
  });
};
export const del = (http, params) => {
  return raw.del(http, getAppInstallationUrl(params));
};
export const getForOrganization = (http, params) => {
  return raw.get(http, getBaseUrlForOrgInstallations(params), {
    params: _objectSpread(_objectSpread({}, normalizeSpaceId(normalizeSelect(params.query))), {}, {
      'sys.organization.sys.id[in]': params.organizationId
    })
  });
};
