// contentful-management/dist/es-modules/adapters/REST/endpoints/app-definition.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import * as raw from './raw';
import copy from 'fast-copy';
import { normalizeSelect, normalizeSpaceId } from './utils';
const getBaseUrl = params => `/organizations/${params.organizationId}/app_definitions`;
export const getAppDefinitionUrl = params => getBaseUrl(params) + `/${params.appDefinitionId}`;
const getBaseUrlForOrgInstallations = params => `/app_definitions/${params.appDefinitionId}/app_installations`;
export const get = (http, params) => {
  return raw.get(http, getAppDefinitionUrl(params), {
    params: normalizeSelect(params.query)
  });
};
export const getMany = (http, params) => {
  return raw.get(http, getBaseUrl(params), {
    params: params.query
  });
};
export const create = (http, params, rawData) => {
  const data = copy(rawData);
  return raw.post(http, getBaseUrl(params), data);
};
export const update = (http, params, rawData, headers) => {
  var _rawData$sys$version;
  const data = copy(rawData);
  delete data.sys;
  return raw.put(http, getAppDefinitionUrl(params), data, {
    headers: _objectSpread({
      'X-Contentful-Version': (_rawData$sys$version = rawData.sys.version) !== null && _rawData$sys$version !== void 0 ? _rawData$sys$version : 0
    }, headers)
  });
};
export const del = (http, params) => {
  return raw.del(http, getAppDefinitionUrl(params));
};
export const getInstallationsForOrg = (http, params) => {
  return raw.get(http, getBaseUrlForOrgInstallations(params), {
    params: _objectSpread(_objectSpread({}, normalizeSpaceId(normalizeSelect(params.query))), {}, {
      'sys.organization.sys.id[in]': params.organizationId
    })
  });
};
