// contentful-management/dist/es-modules/adapters/REST/endpoints/team-space-membership.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import copy from 'fast-copy';
import * as raw from './raw';
const getBaseUrl = params => `/spaces/${params.spaceId}/team_space_memberships`;
const getEntityUrl = params => `${getBaseUrl(params)}/${params.teamSpaceMembershipId}`;
export const get = (http, params) => raw.get(http, getEntityUrl(params));
export const getMany = (http, params) => raw.get(http, getBaseUrl(params), {
  params: params.query
});
export const getForOrganization = (http, params) => {
  return raw.get(http, `/organizations/${params.organizationId}/team_space_memberships/${params.teamSpaceMembershipId}`);
};
export const getManyForOrganization = (http, params) => {
  const query = params.query || {};
  if (params.teamId) {
    query['sys.team.sys.id'] = params.teamId;
  }
  return raw.get(http, `/organizations/${params.organizationId}/team_space_memberships`, {
    params: params.query
  });
};
export const create = (http, params, rawData, headers) => {
  return raw.post(http, getBaseUrl(params), rawData, {
    headers: _objectSpread({
      'x-contentful-team': params.teamId
    }, headers)
  });
};
export const update = (http, params, rawData, headers) => {
  const data = copy(rawData);
  delete data.sys;
  return raw.put(http, getEntityUrl(params), data, {
    headers: _objectSpread({
      'X-Contentful-Version': rawData.sys.version || 0,
      'x-contentful-team': rawData.sys.team.sys.id
    }, headers)
  });
};
export const del = (http, params) => {
  return raw.del(http, getEntityUrl(params));
};
