// contentful-management/dist/es-modules/entities/workflow-definition.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import copy from 'fast-copy';
import { wrapCollection } from '../common-utils';
import enhanceWithMethods from '../enhance-with-methods';

/* Workflow Step Permission */

export let WorkflowStepPermissionType = /*#__PURE__*/function (WorkflowStepPermissionType) {
  WorkflowStepPermissionType["EntityPermission"] = "entity_permission";
  WorkflowStepPermissionType["WorkflowPermission"] = "workflow_permission";
  return WorkflowStepPermissionType;
}({});
export let WorkflowStepPermissionAction = /*#__PURE__*/function (WorkflowStepPermissionAction) {
  WorkflowStepPermissionAction["Edit"] = "edit";
  WorkflowStepPermissionAction["Publish"] = "publish";
  return WorkflowStepPermissionAction;
}({});
export let WorkflowStepPermissionEffect = /*#__PURE__*/function (WorkflowStepPermissionEffect) {
  WorkflowStepPermissionEffect["Allow"] = "allow";
  WorkflowStepPermissionEffect["Deny"] = "deny";
  return WorkflowStepPermissionEffect;
}({});
/* Workflow Step Action */
export let WorkflowStepActionType = /*#__PURE__*/function (WorkflowStepActionType) {
  WorkflowStepActionType["App"] = "app";
  WorkflowStepActionType["Email"] = "email";
  WorkflowStepActionType["Task"] = "task";
  return WorkflowStepActionType;
}({});

/* Workflow Step */

/* Workflow Definition */

/**
 * @private
 */
export default function createWorkflowDefinitionApi(makeRequest) {
  const getParams = workflowDefinition => ({
    spaceId: workflowDefinition.sys.space.sys.id,
    environmentId: workflowDefinition.sys.environment.sys.id,
    workflowDefinitionId: workflowDefinition.sys.id
  });
  return {
    update: function update() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'WorkflowDefinition',
        action: 'update',
        params: getParams(raw),
        payload: raw
      }).then(data => wrapWorkflowDefinition(makeRequest, data));
    },
    delete: function _delete() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'WorkflowDefinition',
        action: 'delete',
        params: _objectSpread(_objectSpread({}, getParams(raw)), {}, {
          version: raw.sys.version
        })
      }).then(() => {
        // noop
      });
    }
  };
}

/**
 * @private
 */
export function wrapWorkflowDefinition(makeRequest, data) {
  const workflowDefinition = toPlainObject(copy(data));
  const workflowDefinitionWithMethods = enhanceWithMethods(workflowDefinition, createWorkflowDefinitionApi(makeRequest));
  return freezeSys(workflowDefinitionWithMethods);
}

/**
 * @private
 */
export const wrapWorkflowDefinitionCollection = wrapCollection(wrapWorkflowDefinition);
