// @contentful/app-scripts/lib/upsert-actions/get-cli-args.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCreateAppActionsArgs = getCreateAppActionsArgs;
const chalk_1 = __importDefault(require("chalk"));
const ora_1 = __importDefault(require("ora"));
const get_app_info_1 = require("../get-app-info");
const validate_arguments_1 = require("../validate-arguments");
const constants_1 = require("../constants");
const requiredOptions = {
    organizationId: '--organization-id',
    definitionId: '--definition-id',
    token: '--token',
};
async function getCreateAppActionsArgs(settings) {
    const validateSpinner = (0, ora_1.default)('Validating your input').start();
    try {
        (0, validate_arguments_1.validateArguments)(requiredOptions, settings, 'upsert-actions');
        const appInfo = await (0, get_app_info_1.getAppInfo)(settings);
        return {
            host: settings.host || 'api.contentful.com',
            manifestFile: settings.manifestFile || constants_1.DEFAULT_APP_MANIFEST_PATH,
            accessToken: settings.token,
            appDefinitionId: appInfo.definition.value,
            organizationId: appInfo.organization.value,
        };
    }
    catch (err) {
        console.log(`
			${chalk_1.default.red('Validation failed')}
			${err.message}
		`);
        // eslint-disable-next-line no-process-exit
        process.exit(1);
    }
    finally {
        validateSpinner.stop();
    }
}
