// @contentful/app-scripts/lib/upsert-actions/make-cma-payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAppActionCMAPayload = makeAppActionCMAPayload;
function makeAppActionCMAPayload(action) {
    const baseProps = {
        name: action.name,
        ...(action.description && { description: action.description }),
        ...(action.id && { id: action.id }),
    };
    const additionalPropsByType = action.type === 'function-invocation'
        ? {
            type: action.type,
            function: {
                sys: {
                    id: action.functionId,
                    linkType: 'Function',
                    type: 'Link',
                },
            },
        }
        : {
            type: action.type,
            url: action.url,
        };
    const additionalPropsByCategory = action.category === 'Custom'
        ? {
            category: action.category,
            parameters: action.parameters,
        }
        : {
            category: action.category,
        };
    const payload = {
        ...baseProps,
        ...additionalPropsByType,
        ...additionalPropsByCategory,
    };
    return payload;
}
