// @contentful/app-scripts/lib/upsert-actions/prompt-interactive-args.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.promptCreateAppAction = promptCreateAppAction;
const inquirer_1 = require("inquirer");
const constants_1 = require("../constants");
const get_app_info_1 = require("../get-app-info");
async function promptCreateAppAction(options) {
    const { manifestFile, host } = options;
    const prompts = [];
    if (manifestFile === undefined) {
        prompts.push({
            type: 'input',
            name: 'manifestFile',
            message: `Path to your Contentful app manifest file:`,
            default: constants_1.DEFAULT_APP_MANIFEST_PATH,
        });
    }
    if (!host) {
        prompts.push({
            name: 'host',
            message: `Contentful CMA endpoint URL:`,
            default: 'api.contentful.com',
        });
    }
    const { host: interactiveHost, manifestFile: interactiveManifest } = await (0, inquirer_1.prompt)(prompts);
    const hostValue = host || interactiveHost;
    const manifestFileValue = manifestFile || interactiveManifest;
    const appInfo = await (0, get_app_info_1.getAppInfo)({ ...options, host: hostValue });
    return {
        host: hostValue,
        manifestFile: manifestFileValue,
        appDefinitionId: appInfo.definition.value,
        accessToken: appInfo.accessToken,
        organizationId: appInfo.organization.value,
    };
}
