// @contentful/f36-utils/dist/index.js
'use strict';

var u = require('react');
var reactDom = require('react-dom');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var u__default = /*#__PURE__*/_interopDefault(u);

function y({children:t,container:r}){let e=u.useRef(r),o=u.useRef(document.createElement("div"));return u.useEffect(()=>{e.current||(e.current=document.body);let n=o.current;return n.setAttribute("data-cf-ui-portal",""),e.current.appendChild(n),()=>{e.current&&e.current.removeChild(n);}},[]),o.current?reactDom.createPortal(t,o.current):null}var g=t=>{let{ref:r,keys:e,event:o="keydown"}=t,n=u.useRef(document),c=u.useCallback(s=>{Object.prototype.hasOwnProperty.call(e,s.key)&&e[s.key](s);},[e]);u.useEffect(()=>(r&&r.current&&(n.current=r.current),n.current.addEventListener(o,c),()=>{n.current.removeEventListener(o,c);}),[o,c,r]);};function h(t,r){let e={before:"",match:"",after:""},o=r.replace(/[.*+?^${}()|[\]\\]/g,"\\$&"),n=new RegExp(`(?<before>.*?)(?<match>${o})(?<after>.*)`,"i"),c=t.match(n);return c&&(e.before=c.groups.before,e.match=c.groups.match,e.after=c.groups.after),e}function R(t,r=1){let e=parseInt(t.slice(1,3),16),o=parseInt(t.slice(3,5),16),n=parseInt(t.slice(5,7),16);return `rgba(${e}, ${o}, ${n}, ${r})`}function E(t){return u.isValidElement(t)&&!!t.props.children}function p(t){let r="";return u.Children.toArray(t).map(e=>{u.isValidElement(e)&&E(e)?r+=p(e.props.children):(typeof e=="string"||typeof e=="number")&&(r=`${r}${e}`);}),r}var f=u__default.default.createContext("low"),v=()=>u__default.default.useContext(f),D=f.Provider;

exports.DensityProvider = D;
exports.Portal = y;
exports.getStringMatch = h;
exports.getTextFromChildren = p;
exports.hexToRGBA = R;
exports.useDensity = v;
exports.useKeyboard = g;
