// @contentful/app-scripts/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLocations = exports.upsertActions = exports.generateFunction = exports.buildFunctions = exports.install = exports.feedback = exports.track = exports.open = exports.cleanup = exports.activate = exports.upload = exports.createAppDefinition = void 0;
var create_app_definition_1 = require("./create-app-definition");
Object.defineProperty(exports, "createAppDefinition", { enumerable: true, get: function () { return create_app_definition_1.createAppDefinition; } });
var upload_1 = require("./upload");
Object.defineProperty(exports, "upload", { enumerable: true, get: function () { return upload_1.upload; } });
var activate_1 = require("./activate");
Object.defineProperty(exports, "activate", { enumerable: true, get: function () { return activate_1.activate; } });
var clean_up_1 = require("./clean-up");
Object.defineProperty(exports, "cleanup", { enumerable: true, get: function () { return clean_up_1.cleanup; } });
var open_1 = require("./open");
Object.defineProperty(exports, "open", { enumerable: true, get: function () { return open_1.open; } });
var analytics_1 = require("./analytics");
Object.defineProperty(exports, "track", { enumerable: true, get: function () { return analytics_1.track; } });
var feedback_1 = require("./feedback");
Object.defineProperty(exports, "feedback", { enumerable: true, get: function () { return feedback_1.feedback; } });
var install_1 = require("./install");
Object.defineProperty(exports, "install", { enumerable: true, get: function () { return install_1.install; } });
var build_functions_1 = require("./build-functions");
Object.defineProperty(exports, "buildFunctions", { enumerable: true, get: function () { return build_functions_1.buildFunctions; } });
var generate_function_1 = require("./generate-function");
Object.defineProperty(exports, "generateFunction", { enumerable: true, get: function () { return generate_function_1.generateFunction; } });
var upsert_actions_1 = require("./upsert-actions");
Object.defineProperty(exports, "upsertActions", { enumerable: true, get: function () { return upsert_actions_1.upsertActions; } });
var add_locations_1 = require("./add-locations");
Object.defineProperty(exports, "addLocations", { enumerable: true, get: function () { return add_locations_1.addLocations; } });
